File: data/tutorial_wp/ex-3-fpower-answer.c
/* run.config
   DEPS: @PTEST_DEPS@ @PTEST_DIR@/@PTEST_NAME@.@PTEST_NUMBER@.session@PTEST_CONFIG@/interactive/*.v
   STDOPT:+"-wp-no-rte -wp-prover alt-ergo,coq -wp-session @PTEST_DIR@/@PTEST_NAME@.@PTEST_NUMBER@.session@PTEST_CONFIG@"
*/

#include <limits.h>

/*@
  inductive is_power(integer x, integer n, integer r) {
  case zero: \forall integer x ; is_power(x, 0, 1) ;
  case N: \forall integer x, n, r ; n > 0 ==> is_power(x, n-1, r) ==> is_power(x, n, r*x) ;
  }
*/

/*@
  axiomatic Power{
    axiom power_even:
        \forall integer x, n, r ;
        n >= 0 ==> is_power(x * x, n, r) ==> is_power(x, n * 2, r) ;

    axiom power_odd:
        \forall integer x, n, rp ;
        n >= 0 ==> is_power(x * x, n, rp) ==> is_power(x, 2 * n + 1, x * rp) ;
    axiom monotonic:
        \forall integer x, n1, n2, r1, r2;
        0 <= n1 <= n2 ==> is_power(x, n1, r1) && is_power(x, n2, r2) ==> r1 <= r2;
  }
*/

/*@
  requires 0 <= n < INT_MAX ;
  requires \exists integer m; is_power(x, n, m) && INT_MIN <= m <= INT_MAX;
  assigns \nothing ;
  ensures is_power(x, n, \result);
*/
int power(int x, int n){
  int r = 1 ;

  /*@
    loop invariant 1 <= i <= n+1 ;
    loop invariant is_power(x, i-1, r) ;
    loop assigns i, r ;
    loop variant n - i ;
  */
  for(int i = 1 ; i <= n ; ++i){
    r *= x ;
  }

  return r ;
}

/*@
  assigns \nothing ;
  ensures is_power(x, n, \result);
*/
unsigned fast_power(unsigned x, unsigned n){
  //@ ghost unsigned n_pre = n;
  unsigned r = 1 ;
  unsigned p = x ;

  /*@
    loop invariant 0 <= n ;
    loop invariant \forall integer v ; is_power(p, n, v) ==> is_power(x, n_pre, r * v) ;
    loop assigns n, r, p ;
    loop variant n ;
  */
  while(n > 0){
    if(n % 2 == 1){
        //@ assert is_power(p, n, r);
        r = r * p ;
    }
    p *= p ;
    n /= 2 ;
  }
  //@ assert is_power(p, n, 1) ;

  return r ;
}

File: RQ1_result/migration_failure/ex-3-fpower-answer.rs
#[allow(unused_imports)]
use vstd::math::*;
use vstd::prelude::*;
use vstd::slice::*;
verus! {

pub uninterp spec fn is_power(x: int, n: int, r: int) -> bool;

#[verifier::external_body]
pub broadcast proof fn is_power_inductive()
    ensures
        forall|x: int| (is_power(x as int, 0 as int, 1 as int)),
        forall|x: int, n: int, r: int|
            (n > 0) as int != 0 ==> (((is_power(x as int, n - 1 as int, r as int))) as int != 0
                ==> ((is_power(x as int, n as int, r * x as int))) as int != 0) as int != 0,
{
}

#[verifier::external_body]
pub broadcast proof fn power_even()
    ensures
        forall|x: int, n: int, r: int|
            (n >= 0) as int != 0 ==> (((is_power(x * x as int, n as int, r as int))) as int != 0
                ==> ((is_power(x as int, n * 2 as int, r as int))) as int != 0) as int != 0,
{
}

#[verifier::external_body]
pub broadcast proof fn power_odd()
    ensures
        forall|x: int, n: int, rp: int|
            (n >= 0) as int != 0 ==> (((is_power(x * x as int, n as int, rp as int))) as int != 0
                ==> ((is_power(x as int, 2 * n + 1 as int, x * rp as int))) as int != 0) as int
                != 0,
{
}

#[verifier::external_body]
pub broadcast proof fn monotonic()
    ensures
        forall|x: int, n1: int, n2: int, r1: int, r2: int|
            (0 <= n1 <= n2) as int != 0 ==> (((is_power(x as int, n1 as int, r1 as int)) && (
            is_power(x as int, n2 as int, r2 as int))) as int != 0 ==> (r1 <= r2) as int
                != 0) as int != 0,
{
}

#[verifier::loop_isolation(false)]
#[verifier::exec_allows_no_decreases_clause]
fn power(x: i32, n: i32) -> (result: i32)
    requires
        0 <= n < i32::MAX,
        exists|m: int| (is_power(x as int, n as int, m as int)) && i32::MIN <= m <= i32::MAX,
    ensures
        (is_power(x as int, n as int, result as int)),
{
    broadcast use is_power_inductive, power_even, power_odd, monotonic;

    let mut r: i32 = 1;
    let mut i: i32 = 1;
    while i <= n
        invariant
            1 <= i <= n + 1,
            (is_power(x as int, i - 1 as int, r as int)),
        decreases n - i,
    {
        r *= x;
        i += 1;
    }
    r
}

#[verifier::loop_isolation(false)]
#[verifier::exec_allows_no_decreases_clause]
fn fast_power(x: u32, n: u32) -> (result: u32)
    ensures
        (is_power(x as int, n as int, result as int)),
{
    broadcast use is_power_inductive, power_even, power_odd, monotonic;

    let n_pre: u32 = n;
    let mut r: u32 = 1;
    let mut p: u32 = x;
    let mut n: u32 = n;
    while n > 0
        invariant
            0 <= n,
            forall|v: int|
                ((is_power(p as int, n as int, v as int))) as int != 0 ==> ((is_power(
                    x as int,
                    n_pre as int,
                    r * v as int,
                ))) as int != 0,
        decreases n,
    {
        if n % 2 == 1 {
            assert((is_power(p as int, n as int, r as int)));
            r = r * p;
        }
        p *= p;
        n /= 2;
    }
    assert((is_power(p as int, n as int, 1 as int)));
    r
}

fn main() {
}

} // verus!
